### ISSUE TITLE
Warning alerts are never sent in DTLS

### ISSUE THREAD

**Issue by LyudmilaBalueva** [openssl/openssl#8998]
Function dtls1_dispatch_alert in d1_msg.c only calls BIO_flush for fatal alerts:
'''
        if (s->s3->send_alert[0] == SSL3_AL_FATAL)
            (void)BIO_flush(s->wbio);
'''
As a result, outstanding warning alerts never go to the peer. 

We're using memory BIO, and tested this for NO_RENEGOTIATION alert. The alert was generated, corresponding information callback was issued, but it was not possible to get the alert record from OpenSSL. When we corrected it locally (added BIO_flush call for all alerts), the issue has gone, alert reached the remote side.

Similar issues was earlier found and fixed in ssl3_dispatch_alert in s3_msg.c, it now contains BIO_flush for all alerts: https://github.com/openssl/openssl/pull/3432/commits/39e2c3c073d191081af0f30b869cec24a0ba82ea 

Seems the same fix shall be mapped into dtls1_dispatch_alert function.

We're using our own build of OpenSSL 1.1.1a, the protocol is DTLS 1.2.

Thanks.

**Comment by agnosticdev** [comment]
> When we corrected it locally (added BIO_flush call for all alerts), the issue has gone, alert reached the remote side.

When you corrected the issue locally, was that removing the conditional for:
```c
if (s->s3.send_alert[0] == SSL3_AL_FATAL)
```
and just having `BIO_flush` without the conditional?

**Comment by LyudmilaBalueva** [comment]
Yes

**Comment by agnosticdev** [comment]
Thanks.  During your debugging did you find anything that led you to see why `s->s3.send_alert[0]` was not being equated to `SSL3_AL_FATAL` ?

**Comment by LyudmilaBalueva** [comment]
"No renegotiation" alert is always warning isn't it?
We could only reproduce one warning alert, it's no_renegotiation. 

Or, you mean that s->s3.send_alert[0] does not point to alert level?

**Comment by agnosticdev** [comment]
Yes, sorry.  Why `s->s3.send_alert[0]` does not point to alert level?

**Comment by LyudmilaBalueva** [comment]
I have not debugged what s->s3.send_alert[0] actually points to. From the condition it looks like it should be the alert level, and we only see the problem with warning alert.

**Comment by mattcaswell** [comment]
Looking at the code, I can't see a good reason for this conditional, so it looks reasonable to me to just delete it (keeping the BIO_flush).

Probably this is a good first issue for someone.

**Comment by agnosticdev** [comment]
Thank you @mattcaswell 
@LyudmilaBalueva do you want to delete the conditional for:
```c
if (s->s3.send_alert[0] == SSL3_AL_FATAL)
```

**Comment by LyudmilaBalueva** [comment]
I want to have it deleted, but not by me. 
Maybe someone else will do that?

**Comment by agnosticdev** [comment]
Sure, I can put in a PR for the remove of this conditional.

**Comment by agnosticdev** [comment]
PR is up: https://github.com/openssl/openssl/pull/9041

**Comment by mattcaswell** [comment]
Fixed via #9041.
